boa/src/builtins/number/mod.rs

```diff
@@ -405,6 +411,132 @@ impl Number {
         Ok(Self::to_number(this))
     }
 
+    /// Builtin javascript 'parseInt(str, radix)' function.
+    ///
+    /// Parses the given string as an integer using the given radix as a base.
+    ///
+    /// An argument of type Number (i.e. Integer or Rational) is also accepted in place of string.
+    ///
+    /// The radix must be an integer in the range [2, 36] inclusive.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-parseint-string-radix
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt
+    pub(crate) fn parse_int(
+        _this: &mut Value,
+        args: &[Value],
+        _ctx: &mut Interpreter,
+    ) -> ResultValue {
+        if args.len() > PARSE_INT_MAX_ARG_COUNT {
+            // Too many arguments.
+            return Err(Value::undefined());
+        }
+
+        if let (Some(val), r) = (args.get(0), args.get(1)) {
+            let mut radix = if let Some(rx) = r {
+                if let ValueData::Integer(i) = rx.data() {
+                    *i as u32
+                } else {
+                    // Handling a second argument that isn't an integer but was provided so cannot be defaulted.
+                    return Ok(Value::from(f64::NAN));
+                }
+            } else {
+                // No second argument provided therefore radix is unknown
+                0
+            };
+
+            match val.data() {
+                ValueData::String(s) => {
+                    // Attempt to infer radix from given string.
+
+                    if radix == 0 {
+                        if s.starts_with("0x") || s.starts_with("0X") {
+                            if let Ok(i) = i32::from_str_radix(&s[2..], 16) {
+                                return Ok(Value::integer(i));
+                            } else {
+                                // String can't be parsed.
+                                return Ok(Value::from(f64::NAN));
+                            }
+                        } else {
+                            radix = 10
+                        };
+                    }
+
+                    if let Ok(i) = i32::from_str_radix(s, radix) {
+                        Ok(Value::integer(i))
+                    } else {
+                        // String can't be parsed.
+                        Ok(Value::from(f64::NAN))
+                    }
+                }
+                ValueData::Integer(i) => Ok(Value::integer(*i)),
+                ValueData::Rational(f) => Ok(Value::integer(*f as i32)),
+                _ => {
+                    // Wrong argument type to parseInt.
+                    Ok(Value::from(f64::NAN))
+                }
+            }
+        } else {
+            // Not enough arguments to parseInt.
+            Err(Value::undefined())
```

Comment:
Note that if no arguments are passed, then the first argument is `undefined`, and therefore, it should return `NaN`, without throwing an exception (`Err` throws an exception with the given value).

Reply:
Ah ok, I wasn't sure how to handle too many/few arguments, I've made the changes in the latest commit